### PR TITLE
Add without aggregator modifier.

### PR DESCRIPTION
**Pull request by brian-brazil**
This has the advantage that the user doesn't need
to list all labels they want to keep (as with "by")
but without having to worry about inconsistent labels
as when there's only one time series (as with "keeping_common").
    
Almost all aggregation should use this rather than the existing
two options as it's much less error prone and easier to maintain
due to not having to always add in "job" plus whatever other common
job-level labels you have like "region".

@fabxc In terms of aggregation, this should make us feature complete.

### CODE REVIEW

File: promql/engine.go [hunk 1]
```diff
@@ -1039,12 +1039,26 @@ type groupedAggregation struct {
...
-func (ev *evaluator) aggregation(op itemType, grouping model.LabelNames, keepExtra bool, vec vector) vector {
+func (ev *evaluator) aggregation(op itemType, grouping model.LabelNames, without bool, keepExtra bool, vec vector) vector {
 
 	result := map[uint64]*groupedAggregation{}
 
 	for _, sample := range vec {
-		groupingKey := model.SignatureForLabels(sample.Metric.Metric, grouping...)
+		var withoutMetric metric.Metric
+		if without {
+			withoutMetric = *sample.Metric.Copy()
```
Comment (fabxc) [inline promql/engine.go]: For `by` we build the metric in the if condition above. Doing it here for without makes things inconsistent and hard to read.

Also `metric.Metric` is already the copy-on-write version. So this line can essentially be dropped.

Comment (brian-brazil) [inline promql/engine.go]: The problem is we need this version in two places, so I didn't want to duplicate the code.

Dropped the line.

File: promql/engine.go [hunk 2]
```diff
@@ -1039,12 +1039,25 @@ type groupedAggregation struct {
...
 
 // aggregation evaluates an aggregation operation on a vector.
-func (ev *evaluator) aggregation(op itemType, grouping model.LabelNames, keepExtra bool, vec vector) vector {
+func (ev *evaluator) aggregation(op itemType, grouping model.LabelNames, without bool, keepExtra bool, vec vector) vector {
 
 	result := map[uint64]*groupedAggregation{}
 
 	for _, sample := range vec {
-		groupingKey := model.SignatureForLabels(sample.Metric.Metric, grouping...)
+    withoutMetric := sample.Metric
```
Comment (fabxc) [inline promql/engine.go]: gofmt

File: promql/parse.go [hunk 3]
```diff
@@ -732,11 +732,14 @@ func (p *parser) aggrExpr() *AggregateExpr {
 		p.errorf("expected aggregation operator but got %s", agop)
 	}
 	var grouping model.LabelNames
-	var keepExtra bool
+	var keepExtra, without bool
 
 	modifiersFirst := false
 
-	if p.peek().typ == itemBy {
+	if p.peek().typ == itemBy || p.peek().typ == itemWithout {
```
Comment (fabxc) [inline promql/parse.go]: We can reduce noise a bit by doing the following. Similar below.

```go
if t := p.peek().typ; t == itemBy || t == itemWithout{
    // ...
   if t == itemWithout {
       // ...
```

Comment (brian-brazil) [inline promql/parse.go]: Done

File: promql/engine.go (hunk 2 above)
Comment (brian-brazil) [inline promql/engine.go]: Done

**Comment by fabxc** [comment]
:+1:

**Comment by matthiasr** [comment]
I realize I'm late to the party, but if this does what I think it does I would like to propose calling it `over` instead of `without`. The way to read that would be "sum over (dimensions)". To me, removing labels is a side effect of aggregations, not what they are about at their core. Additionally, " sum without" suggests that something is not being included in the sum, which is not the case.

**Comment by matthiasr** [comment]
(just putting this out there before this is in a release)

**Comment by brian-brazil** [comment]
"over" sounds more like what "by" does by default. I'd consider removing labels to be a key part of aggregation.

**Pull request merged**